File: ml_tools/ui/service/control.py
# Keep in sync with ml_tools/ui/server/service/control.go

import json

from ..protocol import Message, Protocol
from . import message_type


class Control:
  message_type_range = (100, 199)

  def __init__(self, protocol: Protocol):
    self.protocol = protocol
    self.map = {}

  def on_message(self, message: Message):
    if message.Type == message_type.set_value:
      data = json.loads(message.Data)
      self.map[data['key']] = data['value']

  def get_value(self, key: str) -> str:
    return self.map.get(key)

File: ml_tools/ui/server/service/control.go
package service

// Keep in sync with ml_tools/ui/service/control.py

import (
	"encoding/json"
	"ml_tools/server/protocol"
)

type Control struct {
	pw protocol.ReadWriter
}

func (c *Control) GetMessageTypeRange() (min, max int) {
	return 100, 199
}

type setValueModel struct {
	Key   string `json:"key"`
	Value string `json:"value"`
}

func (c *Control) SetValue(key, value string) error {
	data, err := json.Marshal(setValueModel{
		Key:   key,
		Value: value,
	})
	if err != nil {
		return nil
	}

	return c.pw.Write(protocol.Message{
		Type: SetValue,
		Data: data,
	})
}

func (c *Control) OnMessage(message protocol.Message) {}
